config/configmapprovider/provider.go

```diff
@@ -20,42 +20,77 @@ import (
 	"go.opentelemetry.io/collector/config"
 )
 
-// Provider is an interface that helps providing configuration's parser.
-// Implementations may load the parser from a file, a database or any other source.
+// Provider is an interface that helps to retrieve a config map and watch for any
+// changes to the config map. Implementations may load the config from a file,
+// a database or any other source.
 type Provider interface {
 	// Retrieve goes to the configuration source and retrieves the selected data which
 	// contains the value to be injected in the configuration and the corresponding watcher that
 	// will be used to monitor for updates of the retrieved value.
-	Retrieve(ctx context.Context) (Retrieved, error)
+	//
+	// onChange callback is called when the config changes. onChange may be called from
+	// a different go routine. After onChange is called Retrieved.Get should be called
+	// to get the new config. See description of Retrieved for more details.
+	// onChange may be nil, which indicates that the caller is not interested in
+	// knowing about the changes.
+	//
+	// If ctx is cancelled should return immediately with an error.
+	// Should never be called concurrently with itself or with Shutdown.
+	Retrieve(ctx context.Context, onChange func(*ChangeEvent)) (Retrieved, error)
 
-	// Close signals that the configuration for which it was used to retrieve values is no longer in use
-	// and the object should close and release any watchers that it may have created.
-	// This method must be called when the service ends, either in case of success or error.
-	Close(ctx context.Context) error
+	// Shutdown signals that the configuration for which this Provider was used to
+	// retrieve values is no longer in use and the Provider should close and release
+	// any resources that it may have created.
+	//
+	// This method must be called when the Collector service ends, either in case of
+	// success or error. Retrieve cannot be called after Shutdown.
+	//
+	// Should never be called concurrently with itself or with Retrieve.
+	// If ctx is cancelled should return immediately with an error.
+	Shutdown(ctx context.Context) error
 }
 
-// Retrieved holds the result of a call to the Retrieve method of a Session object.
+// Retrieved holds the result of a call to the Retrieve method of a Provider object.
+//
+// The typical usage is the following:
+//
+//		r := mapProvider.Retrieve()
+//		r.Get()
+//		// wait for onChange() to be called.
+//		r.Close()
```

Comment:
I think after calling "Close" no more "onChange" should happen correct?

```
//		r := mapProvider.Retrieve()
//		r.Get()
//		// wait for onChange() to be called.
//		r.Close()
//		r = mapProvider.Retrieve()
//		r.Get()
//		// wait for onChange() to be called.
//		r.Close()
```

It simplifies if "Retrieved" is immutable for "Get". Otherwise you need to protect between a normal caller calling Get and the watcher coming and updating the value.

Reply:
Good catch, docs are incorrect. The cycle is Retrieve/Get/wait/Close.